File: simulator/src/main/java/it/unibo/scafi/simulation/gui/view/GuiNode.java
package it.unibo.scafi.simulation.gui.view;

import it.unibo.scafi.simulation.gui.model.Node;
import it.unibo.scafi.simulation.gui.utility.Utils;

import javax.swing.*;
import java.awt.*;
import java.beans.PropertyVetoException;
import java.nio.file.Paths;

/**
 * This class represents a
 * graphics representation of Node
 *
 * Created by chiara on 01/06/16.
 */
public class GuiNode extends JInternalFrame {

    //costante per il font del valore visualizzato (valueShow)
    private final Font DEFAULT_FONT = new Font("Arial", Font.BOLD, 14);

    private final JLabel valueShow;
    private final JButton button;
    private NodeInfoPanel infoPanel;    //pannello delle informazioni del nodo
    private boolean isSelected;
    private Node node = null;

    public GuiNode(Node node) {
        //initial settings
        super();
        this.node = node;
        this.valueShow = new JLabel("");
        this.button = new JButton(Utils.createImageIcon("node.png"));
        setSize(Utils.getSizeGuiNode());
        setBorder(null);

        JPanel background = new JPanel(new BorderLayout());
        background.setOpaque(false);    //trasparenza
        setContentPane(background);
        valueShow.setFont(DEFAULT_FONT);

        //pannello in cui visualizzo una singola info (id, export,...)
        JPanel north = new JPanel();
        north.setOpaque(false);
        north.add(valueShow);
        background.add(north, BorderLayout.NORTH);

        //pannello con il bottone con l'icona del nodo
        JPanel pBotton = new JPanel();
        pBotton.setOpaque(false);
        pBotton.add(button);
        button.setBorderPainted(false);
        button.setOpaque(false);
        button.setOpaque(false);
        button.addMouseListener(new GuiNodeListeners(this)); //listener dei movimentis
        button.addMouseMotionListener(new GuiNodeListeners(this));
        background.add(pBotton, BorderLayout.SOUTH);

        revalidate();
        repaint();
        setVisible(true);
        try {
            this.setSelected(false);
        } catch (PropertyVetoException e) {
            e.printStackTrace();
        }
    }

    public Node getNode(){ return this.node; }

    //gestione dimensione nodo
    @Override
    public void setSize(Dimension d) {
        super.setSize(d);
        String nameImg = Paths.get(((ImageIcon) button.getIcon()).getDescription()).getFileName().toString(); //path icona presente sul bottone
        int dividendo = getWidth() < getHeight() ? getWidth() : getHeight();
        //ridimensionamento icone
        button.setIcon(Utils.getScaledImage(nameImg, dividendo / 2, dividendo / 2));
        //ridimensionamento font
        if (d.getHeight() < (Utils.getFrameDimension().getHeight() / 2)){
            this.valueShow.setFont(DEFAULT_FONT.deriveFont(DEFAULT_FONT.getSize()/2));
        } else {
            this.valueShow.setFont(DEFAULT_FONT);
        }
        //ridimensionamento infoPanel
        if (infoPanel != null) {
            infoPanel.setSize(Utils.getMenuSimulationPanelDim());
            infoPanel.setLocation(getLocation().x + getWidth(), getLocation().y);
        }
    }

    //gestione posizionamento nodo e pannello informazioni
    @Override
    public void setLocation(int x, int y) {
        super.setLocation(x,y);
        if(this.infoPanel!=null){
            infoPanel.setLocation(x+getWidth(),y);
        }
    }

    //gestione selezione nodi
    @Override
    public boolean isSelected() {
        return isSelected;
    }

    @Override
    public void setSelected(boolean selected) throws PropertyVetoException {
//non faccio la super perchè se no non riesco a selezionarne più di uno
        if (selected) {
            this.button.setIcon(Utils.getSelectedIcon( button.getIcon()));
            this.valueShow.setForeground(Color.lightGray);
        } else {
            this.button.setIcon(Utils.getNotSelectIcon(button.getIcon()));
            this.valueShow.setForeground(Color.black);
        }
        isSelected = selected;
    }

    //gestione visualizzazione valori
    public void setValueToShow(final String text) {
        this.valueShow.setText("<html>" + text.replaceAll("\n", "<br>"));
    }

    public void setLabelFont(final Font font) {
        this.valueShow.setFont(font);
    }

    public void setLabelColor(final Color color) {
        this.valueShow.setForeground(color);
    }

    public void setInfoPanel(final NodeInfoPanel p) {
        this.infoPanel = p;
    }

    public void setImageButton(final String res) {
        int dividendo = getWidth() < getHeight() ? getWidth() : getHeight();
        button.setIcon(Utils.getScaledImage(res, dividendo / 2, dividendo / 2));
    }

    public void showInfo(final boolean show) {
        infoPanel.setVisible(show);
    }

    public NodeInfoPanel getInfoPanel() {
        return this.infoPanel;
    }
}

File: simulator-gui/src/main/scala/it/unibo/scafi/simulation/gui/view/GuiNode.scala
/*
 * Copyright (C) 2016-2017, Roberto Casadei, Mirko Viroli, and contributors.
 * See the LICENCE.txt file distributed with this work for additional
 * information regarding copyright ownership.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
*/

package it.unibo.scafi.simulation.gui.view

import it.unibo.scafi.simulation.gui.model.Node
import it.unibo.scafi.simulation.gui.utility.Utils
import javax.swing._
import java.awt._
import java.beans.PropertyVetoException
import java.nio.file.Paths

/**
  * This class represents a graphics representation of Node
  */
class GuiNode(val node: Node) extends JInternalFrame {
  final private val NODE_SYMBOL_FONT: Font = new Font("Arial", Font.BOLD, 28)
  final private val DEFAULT_FONT: Font = new Font("Arial", Font.BOLD, 14)
  final private var valueShow: JLabel = null
  final private var button: JButton = null
  final private var valueString: String = null
  private var infoPanel: NodeInfoPanel = null

  this.valueShow = new JLabel("")
  this.valueShow.setBackground(null)
  this.valueShow.setBorder(null)

  this.button = new JButton(Utils.createImageIcon("node_point.png"))
  this.button.setFont(NODE_SYMBOL_FONT)
  setSize(Utils.getSizeGuiNode)
  setBorder(null)

  val backgroundPanel: JPanel = new JPanel()
  backgroundPanel.setLayout(new BoxLayout(backgroundPanel, BoxLayout.Y_AXIS))
  backgroundPanel.setOpaque(false)
  setContentPane(backgroundPanel)
  valueShow.setFont(DEFAULT_FONT)

  // Panel showing result "above" the node
  val north: JPanel = new JPanel
  north.add(valueShow)
  north.setMaximumSize(new Dimension(100, 22))
  north.setOpaque(false)

  // Panel with button and node icon
  val pBotton: JPanel = new JPanel
  pBotton.setOpaque(false)
  pBotton.add(button)
  pBotton.setMaximumSize(pBotton.getPreferredSize)

  button.setBorder(BorderFactory.createEmptyBorder())
  button.setOpaque(false)
  button.setMaximumSize(new Dimension(10,6))

  button.addMouseListener(new GuiNodeListeners(this))
  button.addMouseMotionListener(new GuiNodeListeners(this))

  backgroundPanel.add(north)
  backgroundPanel.add(pBotton)

  revalidate()
  repaint()

  setVisible(true)

  try {
    this.setSelected(false)
  }
  catch {
    case e: PropertyVetoException => {
      e.printStackTrace()
    }
  }

  override def setSize(d: Dimension) {
    super.setSize(d)
    //val nameImg: String = Paths.get((button.getIcon.asInstanceOf[ImageIcon]).getDescription).getFileName.toString
    val dividendo: Int = if (getWidth < getHeight) getWidth  else getHeight

    //button.setIcon(Utils.getScaledImage(nameImg, dividendo / 2, dividendo / 2))

    if (d.getHeight < (Utils.getFrameDimension.getHeight / 2)) {
      this.valueShow.setFont(DEFAULT_FONT.deriveFont(DEFAULT_FONT.getSize / 2))
    }
    else {
      this.valueShow.setFont(DEFAULT_FONT)
    }

    if (infoPanel != null) {
      infoPanel.setSize(Utils.getMenuSimulationPanelDim)
      infoPanel.setLocation(this.getLocation().x + getWidth, this.getLocation().y)
    }
  }

  override def setLocation(x: Int, y: Int) {
    super.setLocation(x, y)
    if (this.infoPanel != null) {
      infoPanel.setLocation(x + getWidth, y)
    }
  }

  @throws[PropertyVetoException]
  override def setSelected(selected: Boolean) {
    //non faccio la super perchè se no non riesco a selezionarne più di uno
    if (selected) {
      //this.button.setIcon(Utils.getSelectedIcon(button.getIcon))
      this.valueShow.setForeground(Color.lightGray)
    }
    else {
      //this.button.setIcon(Utils.getNotSelectIcon(button.getIcon))
      this.valueShow.setForeground(Color.black)
    }
    isSelected = selected
  }

  def setValueToShow(text: String) {
    this.valueString = text
    this.valueShow.setText("<html>" + text.replaceAll("\n", "<br>"))
  }

  def getValueToShow(): String = this.valueString

  def setLabelFont(font: Font) {
    this.valueShow.setFont(font)
  }

  def setLabelColor(color: Color) {
    this.valueShow.setForeground(color)
  }

  def setInfoPanel(p: NodeInfoPanel) {
    this.infoPanel = p
  }

  def setImageButton(res: String) {
    val dividendo: Int = if (getWidth < getHeight) getWidth
    else getHeight
    //button.setIcon(Utils.getScaledImage(res, dividendo / 2, dividendo / 2))
  }

  def showInfo(show: Boolean) {
    infoPanel.setVisible(show)
  }

  def getInfoPanel: NodeInfoPanel = {
    return this.infoPanel
  }
}
